aiocache/backends/sqlite.py

```diff
@@ -0,0 +1,283 @@
+                self._next_expire_at = when
+                # print("Next cleanup in", when - time())
+    @property
+    def _db_lock(self):
+        if self.__db_lock is None:
+            self.__db_lock = asyncio.Lock()
+        return self.__db_lock
+
+    async def __aenter__(self):
+        await self.open()
+        return self
+
+    async def __aexit__(self, exc_type, exc, tb):
+        await self.close()
+
+    async def open(self):
+        async with self._db_lock:
+            await self._ensure_connected()
+
+    async def close(self):
+        async with self._db_lock:
+            if self._db is not None:
+                await self._db.close()
+                self._db = None
+
+    async def clear_expired(self):
+        async with self._db_lock:
+            await self._ensure_connected()
+            return await self._remove_expired(force=True)
+
+    async def _clear_expired(self, force=False):
+        now = time()
+        if not force:
+            if self._next_expire_at is None or self._next_expire_at >= now + self._cache_expire_interval:
+                # It is too soon to expired data yet
+                return 0
+
+        # Delete expired entries
+        async with self._db.execute("DELETE FROM aiocache WHERE expires_at < ?", (now, )) as cursor:
+            num_deleted = cursor.rowcount
+
+        # Get time of next expiration (So that we don't clear expired entries too soon)
+        async with self._db.execute('SELECT MIN(expires_at) FROM aiocache WHERE expires_at IS NOT NULL') as cursor:
+            result = await cursor.fetchone()
+            self._next_expire_at = result[0] if result else None
+
+        await self._db.commit()
+
+        # print("Expired", num_deleted, "records -- Next cleanup in", "---" if self._next_expire_at is None else self._next_expire_at - now)
+        return num_deleted
+
+    async def _ensure_connected(self):
+        if self._db is None:
+            self._db = await aiosqlite.connect(self._filename)
+            await self._db.execute("""
+                    CREATE TABLE IF NOT EXISTS aiocache (
+                        key TEXT NOT NULL PRIMARY KEY,
+                        value BLOB NOT NULL,
+                        expires_at REAL
+                    );
+            """)
+            await self._db.execute("""
+                    CREATE INDEX IF NOT EXISTS cache_expires_at_idx ON aiocache(expires_at);
+            """)
+            await self._clear_expired(force=True)
+        else:
+            # Already connected
+            await self._clear_expired()  # Good opportunity to invalidate old data
+
+
+
+    async def _get(self, key, encoding="utf-8", _conn=None):
+        return (await self._multi_get([key], encoding=encoding, _conn=_conn))[0]
+
+    async def _gets(self, key, encoding="utf-8", _conn=None):
+        return await self._get(key, encoding=encoding, _conn=_conn)
+
+    async def _multi_get(self, keys, encoding="utf-8", _conn=None):
```

Comment:
ahhh yeah that was a hard thing back then... I tried to move it to the serializer but the cache needed to access to the encoding because implementations like the redis one can only retrieve bytes or strings and need to know the encoding so in the end I decided for this to be a cache level configuration. With just having it in get you can already deal with setting bytes and strings because then when retrieving you specify what you want